aptos-move/framework/supra-stdlib/sources/enumerable_map.move

```diff
@@ -0,0 +1,276 @@
+module supra_std::enumerable_map {
+    use std::error;
+    use std::vector;
+    use aptos_std::table;
+
+    /// Key is already present in the map
+    const EKEY_ALREADY_ADDED: u64 = 1;
+    /// Key is absent in the map
+    const EKEY_ABSENT: u64 = 2;
+    /// Vector is empty
+    const EVECTOR_EMPTY: u64 = 3;
+
+    /// Enumerable Map to store the key value pairs
+    struct EnumerableMap<K : copy + drop, V : store+drop+copy> has store {
+        /// List of all keys
+        list: vector<K>,
+        /// Key mapped to a tuple containing the (position of key in list and value corresponding to the key)
+        map: table::Table<K, Tuple<V>>,
+    }
+
+    /// Tuple to store the position of key in list and value corresponding to the key
+    struct Tuple<V : store+drop+copy> has store, copy, drop {
+        position: u64,
+        value: V,
+    }
+
+    /// Return type
+    struct KeyValue<K : copy + drop, V : store+drop+copy> has store, copy, drop {
+        key: K,
+        value: V,
+    }
+
+    /// To create an empty enum map
+    public fun new_map<K : copy + drop, V : store+drop+copy>(): EnumerableMap<K, V> {
+        return EnumerableMap<K, V> { list: vector::empty<K>(), map: table::new<K, Tuple<V>>() }
+    }
+
+    /// Add Single Key in the Enumerable Map
+    public fun add_value<K : copy+drop, V : store+drop+copy>(map: &mut EnumerableMap<K, V>, key: K, value: V) {
+        assert!(!contains(map, key), error::already_exists(EKEY_ALREADY_ADDED));
+        table::add(&mut map.map, key, Tuple<V> { position: vector::length(&map.list), value });
+        vector::push_back(&mut map.list, key);
+    }
+
+    /// Add Multiple Keys in the Enumerable Map
+    public fun add_value_bulk<K: copy+drop, V : store+drop+copy>(
+        map: &mut EnumerableMap<K, V>,
+        keys: vector<K>,
+        values: vector<V>
+    ): vector<K> {
+        assert!(!vector::is_empty(&values), error::invalid_argument(EVECTOR_EMPTY));
+        let current_key_list_length = vector::length(&map.list);
+        let updated_keys = vector::empty<K>();
+
+        vector::zip_reverse(keys, values, |key, value| {
+            if (!contains(map, key)) {
+                table::add(&mut map.map, key, Tuple<V> { position: current_key_list_length, value });
+                vector::push_back(&mut map.list, key);
+                current_key_list_length = current_key_list_length + 1;
+
+                vector::push_back(&mut updated_keys, key);
+            };
+        });
+
+        return updated_keys
+    }
+
+    /// Update the value of a key thats already present in the Enumerable Map
+    public fun update_value<K: copy+drop, V : store+drop+copy>(
+        map: &mut EnumerableMap<K, V>,
+        key: K,
+        new_value: V
+    ): KeyValue<K, V> {
+        assert!(contains(map, key), error::not_found(EKEY_ABSENT));
+        table::borrow_mut(&mut map.map, key).value = new_value;
+        KeyValue { key, value: new_value }
```

Comment:
@nizam-supraoracles why should it return `KeyValue` with `new_value`? A user who called it would already know the answer. It may be useful if the returned value is `KeyValue { key, old_value}` change the comments also accordingly.

Reply:
will be addressed in scope https://github.com/Entropy-Foundation/smr-moonshot/issues/1425 ticket